README.md
[![Build Status](https://api.travis-ci.org/kolypto/py-password-strength.png?branch=master)](https://travis-ci.org/kolypto/py-password-strength)
[![Pythons](https://img.shields.io/badge/python-2.7%20%7C%203.4%E2%80%933.7%20%7C%20pypy-blue.svg)](.travis.yml)


Password Strength
=================

Password strength and validation.


Tutorial
========


### Uppercase, Numbers, Special Characters

You test your passwords using the Policy object that controls what kind of password is acceptable in your system.

First, create the Policy object and define the rules that apply to passwords in your system:

```python
from password_strength import PasswordPolicy

policy = PasswordPolicy.from_names(
    length=8,  # min length: 8
    uppercase=2,  # need min. 2 uppercase letters
    numbers=2,  # need min. 2 digits
    special=2,  # need min. 2 special characters
    nonletters=2,  # need min. 2 non-letter characters (digits, specials, anything)
)
```

Now, when you have the `PasswordPolicy` object, you can use it to test your passwords,
and it will tell you which tests have failed:

```python
policy.test('ABcd12!')
# -> [Length(8), Special(2)]
```

This tells us that 2 tests have failed: password is not long enough, and it does not have enough special characters.
You can use this information to tell the user what precisely is wrong with their password.

```python
policy.test('ABcd12!@')
# -> []
```

Empty list tells us that this password is alright.

This test, however, enabled uses to use passwords that have a lot of repetition.

### So-Called Entropy Bits

Here's a test that's even better. You don't really need to define complex rules with special characters and stuff.
All you actually need is a password that's long enough, complex enough, and easy to remember
(see [xkcd](https://xkcd.com/936/) and
[Article: Everything We've Been Told About Passwords Is Wrong](https://www.inc.com/thomas-koulopoulos/all-that-advice-about-passwords-turns-out-to-be-to.html)).

So, instead of defining all these rules, let's just require the password to be complex enough.
Entropy bits is something that defines how much variety does your password have. '01111010010011' is long enough,
but has only 2 entropy bits: that's how many bits you need to store its alphabet. However, a password that uses
plenty of characters has more entropy.

```python
policy = PasswordPolicy.from_names(
    entropybits=30  # need a password that has minimum 30 entropy bits (the power of its alphabet)
)

print(policy.test('0123456789'))
# -> []
```

This password is not long enough, or secure enough, but has enough entropy: its vocabulary has 10 different characters.
Put this test together with other requirements to make sure there's no repetition in your passwords.

### Complexity

Entropy bits are important, but difficult to understand. An even better, more intuitive test, is to require
the password to be "complex enough".
Complexity is a number in the range of 0.00..0.99. Good, strong passwords start at 0.66.

Let's first see how different passwords score:

```python
from password_strength import PasswordStats

stats = PasswordStats('qwerty123')
print(stats.strength())  #-> Its strength is 0.316

stats = PasswordStats('G00dPassw0rd?!')
print(stats.strength())  #-> Its strength is 0.585

stats = PasswordStats('V3ryG00dPassw0rd?!')
print(stats.strength())  #-> Its strength is 0.767
```

So, 0.66 will be a very good indication of a good password.
Let's implement our policy:

```python
policy = PasswordPolicy.from_names(
    strength=0.66  # need a password that scores at least 0.5 with its strength
)

print(policy.test('V3ryG00dPassw0rd?!'))
# -> []  -- empty list means a good password
```

One good thing about using strength is that it allows users to use national aplhabets with passwords,
which are most secure:

```python
tested_pass = policy.password('Mixed-汉堡包/漢堡包, 汉堡/漢堡')
print(tested_pass.strength())  # -> 0.812 -- very good!
print(tested_pass.test())
#-> []  - good password; it actually scored 0.812
```

Notice how in the last example we use a different approach: `policy.password()` analyzes the password, and then we can
both get its `.strength()`, and `.test()` it according to the current policy.

PasswordPolicy
==============

Perform tests on a password.

Init Policy
-----------

```python
PasswordPolicy(*tests)
```

Init password policy with a list of tests

Alternatively:

```python
PasswordPolicy.from_names(**tests)
```

Init password policy from a dictionary of test definitions.

A test definition is simply:

    { test-name: argument } or { test-name: [arguments] }

Test name is just a lowercased class name.

Example:

    PasswordPolicy.from_names(
        length=8,
        strength=(0.33, 30),
    )


Bundled Tests
-------------

These objects perform individual tests on a password, and report `True` of `False`.


#### tests.EntropyBits(bits)
Test whether the password has >= `bits` entropy bits.

Entropy bits is the number of bits that is required to store the alphabet that's used in a password.
It's a measure of how long is the alphabet.

#### tests.Length(length)
Tests whether password length >= `length`

#### tests.NonLetters(count)
Test whether the password has >= `count` non-letter characters

#### tests.NonLettersLc(count)
Test whether the password has >= `count` non-lowercase characters

#### tests.Numbers(count)
Test whether the password has >= `count` numeric characters

#### tests.Special(count)
Test whether the password has >= `count` special characters

#### tests.Strength(strength, weak_bits=30)
Test whether the password has >= `strength` strength.

A password is evaluated to the strength of 0.333 when it has `weak_bits` entropy bits,
which is considered to be a weak password. Strong passwords start at 0.666.

#### tests.Uppercase(count)
Test whether the password has >= `count` uppercase characters


Testing
-------

After the `PasswordPolicy` is initialized, there are two methods to test:

### PasswordPolicy.password
```python
password(password)
```
Get password stats bound to the tests declared in this policy.

If in addition to tests you need to get statistics (e.g. strength) -- use this object to double calculations.

See [`PasswordStats`](#passwordstats) for more details.

### PasswordPolicy.test
```python
test(password)
```
Perform tests on a password.

Shortcut for: `PasswordPolicy.password(password).test()`.


Custom Tests
------------

ATest is a base class for password tests.

To create a custom test, just subclass it and implement the following methods:

* __init__() that takes configuration arguments
* test(ps) that tests a password, where `ps` is a `PasswordStats` object.


PasswordStats
-------------

PasswordStats allows to calculate statistics on a password.

It considers a password as a unicode string, and all statistics are unicode-based.

Constructor:

```python
from password_strength import PasswordStats
PasswordStats(password)
```


#### PasswordStats.alphabet
Get alphabet: set of used characters

#### PasswordStats.alphabet_cardinality
Get alphabet cardinality: alphabet length

#### PasswordStats.char_categories
Character count per top-level category

The following top-level categories are defined:

- L: letter
- M: Mark
- N: Number
- P: Punctuation
- S: Symbol
- Z: Separator
- C: Other

#### PasswordStats.char_categories_detailed
Character count per unicode category, detailed format.

See: http://www.unicode.org/reports/tr44/#GC_Values_Table

#### PasswordStats.combinations
The number of possible combinations with the current alphabet

#### PasswordStats.count(*categories)
Count characters of the specified classes only

#### PasswordStats.count_except(*categories)
Count characters of all classes except the specified ones

#### PasswordStats.entropy_bits
Get information entropy bits: log2 of the number of possible passwords

https://en.wikipedia.org/wiki/Password_strength

#### PasswordStats.entropy_density
Get information entropy density factor, ranged {0 .. 1}.

This is ratio of entropy_bits() to max bits a password of this length could have.
E.g. if all characters are unique -- then it's 1.0.
If half of the characters are reused once -- then it's 0.5.

#### PasswordStats.length
Get password length

#### PasswordStats.letters
Count all letters

#### PasswordStats.letters_lowercase
Count lowercase letters

#### PasswordStats.letters_uppercase
Count uppercase letters

#### PasswordStats.numbers
Count numbers

#### PasswordStats.repeated_patterns_length
Detect and return the length of repeated patterns.

You will probably be comparing it with the length of the password itself and ban if it's longer than 10%

#### PasswordStats.sequences_length
Detect and return the length of used sequences:

- Alphabet letters: abcd...
- Keyboard letters: qwerty, etc
- Keyboard special characters in the top row: ~!@#$%^&*()_+
- Numbers: 0123456

#### PasswordStats.special_characters
Count special characters

Special characters is everything that's not a letter or a number

#### PasswordStats.strength(weak_bits=30)
Get password strength as a number normalized to range {0 .. 1}.

Normalization is done in the following fashion:

1. If entropy_bits <= weak_bits   -- linear in range{0.0 .. 0.33} (weak)
2. If entropy_bits <= weak_bits*2 -- almost linear in range{0.33 .. 0.66} (medium)
3. If entropy_bits > weak_bits*3  -- asymptotic towards 1.0 (strong)

#### PasswordStats.test(tests)
Test the password against a list of tests

#### PasswordStats.weakness_factor
Get weakness factor as a float in range {0 .. 1}

This detects the portion of the string that contains:
* repeated patterns
* sequences

E.g. a value of 1.0 means the whole string is weak, and 0.5 means half of the string is weak.

Typical usage:

password_strength = (1 - weakness_factor) * strength



CHANGELOG.md
## 0.0.3 (2019-01-04)
* Python3 support. Finally!


tests/stats-test.py
# -*- coding: utf-8 -*-

import unittest
import six
from password_strength import PasswordStats


class StatsTest(unittest.TestCase):
    """ Test PasswordStats """

    def test_statistics(self):
        self.assertEqual(PasswordStats('123444').alphabet, set('1234'))
        self.assertEqual(PasswordStats(u'!аб!').alphabet, set(u'!аб'))

        self.assertEqual(PasswordStats('123444').alphabet_cardinality, 4)
        self.assertEqual(PasswordStats(u'!аб!').alphabet_cardinality, 3)

        self.assertEqual(
            dict(PasswordStats(u'aAA111!!!!°°°°°      \0').char_categories_detailed),
            {'Ll': 1, 'Lu': 2, 'Nd': 3, 'Po': 4, 'So': 5, 'Zs': 6, 'Cc': 1}
        )

        self.assertEqual(
            dict(PasswordStats(u'aAA111!!!!°°°°°      \0').char_categories),
            {'L': 3, 'N': 3, 'P': 4, 'S': 5, 'Z': 6, 'C': 1}
        )

    def test_count(self):
        s = PasswordStats(u'aAA111!!!!°°°°°      \0')

        self.assertEqual(s.length, 22)
        self.assertEqual(s.letters, 3)
        self.assertEqual(s.letters_lowercase, 1)
        self.assertEqual(s.letters_uppercase, 2)
        self.assertEqual(s.numbers, 3)
        self.assertEqual(s.count('L', 'N'), 3+3)
        self.assertEqual(s.special_characters, 4+5+6+1)

    def test_security(self):
        p2 = '01'
        p8 = '00000001'
        p24 = 'abcdefgh'
        p33 = 'abcdefgh!@'
        p58 = 'abcdefgh!@#$%^&'
        p89 = 'correcthorsebatterystaple'

        p160  = ''.join(six.unichr(n) for n in range(0, 32))
        p384  = ''.join(six.unichr(n) for n in range(0, 64))
        p896  = ''.join(six.unichr(n) for n in range(0, 128))
        p2048 = ''.join(six.unichr(n) for n in range(0, 256))

        self.assertEqual(PasswordStats(p2).combinations,  4)
        self.assertEqual(PasswordStats(p8).combinations,  256)
        self.assertEqual(PasswordStats(p24).combinations, 16777216)

        self.assertAlmostEqual(PasswordStats(   '01').entropy_density,  1.0,  delta=0.01)
        self.assertAlmostEqual(PasswordStats(  '001').entropy_density,  0.63, delta=0.01)
        self.assertAlmostEqual(PasswordStats( '0001').entropy_density,  0.5,  delta=0.01)
        self.assertAlmostEqual(PasswordStats('00001').entropy_density,  0.43, delta=0.01)

        self.assertAlmostEqual(PasswordStats(   p2).entropy_bits, 2.0,     delta=0.01)
        self.assertAlmostEqual(PasswordStats(   p8).entropy_bits, 8.0,     delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p24).entropy_bits, 24.0,    delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p33).entropy_bits, 33.21,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p58).entropy_bits, 58.60,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p89).entropy_bits, 89.62,   delta=0.01)
        self.assertAlmostEqual(PasswordStats( p160).entropy_bits, 160.00,  delta=0.01)
        self.assertAlmostEqual(PasswordStats( p384).entropy_bits, 384.00,  delta=0.01)
        self.assertAlmostEqual(PasswordStats( p896).entropy_bits, 896.00,  delta=0.01)
        self.assertAlmostEqual(PasswordStats(p2048).entropy_bits, 2048.00, delta=0.01)

        self.assertAlmostEqual(PasswordStats(   p2).strength(), 0.02,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(   p8).strength(), 0.08,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p24).strength(), 0.26,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p33).strength(), 0.39,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p58).strength(), 0.70,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(  p89).strength(), 0.88,   delta=0.01)
        self.assertAlmostEqual(PasswordStats( p160).strength(), 0.98,   delta=0.01)
        self.assertAlmostEqual(PasswordStats( p896).strength(), 0.99,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(p2048).strength(), 1.00,   delta=0.01)

        self.assertAlmostEqual(PasswordStats( p2).weakness_factor, 0.0,   delta=0.01)
        self.assertAlmostEqual(PasswordStats( p8).weakness_factor, 0.875, delta=0.01)
        self.assertAlmostEqual(PasswordStats(p24).weakness_factor, 1.0,   delta=0.01)
        self.assertAlmostEqual(PasswordStats(p89).weakness_factor, 0.16,  delta=0.01)

    def test_detectors(self):
        self.assertEqual(PasswordStats('abcabc-1234').repeated_patterns_length, 6)
        self.assertEqual(PasswordStats('abcabcab-1234').repeated_patterns_length, 6)
        self.assertEqual(PasswordStats('abcabcabc-1234').repeated_patterns_length, 9)

        self.assertEqual(PasswordStats('qazwsx').sequences_length, 0)
        self.assertEqual(PasswordStats('qw...').sequences_length, 0)  # Does not detect 2-character sequences
        self.assertEqual(PasswordStats('qwe...').sequences_length, 3)
        self.assertEqual(PasswordStats('qwerty...').sequences_length, 6)
        self.assertEqual(PasswordStats('ZZqwertyZZ1234...').sequences_length, 10)


tests/policy-test.py
import unittest
from password_strength import PasswordPolicy, tests


class PolicyTest(unittest.TestCase):
    """ Test: PasswordPolicy """
    longMessage = True

    def test(self):
        policy = PasswordPolicy.from_names(
            length=8,
            uppercase=2,
            numbers=2,
            special=2,
            nonletters=2,
            nonletterslc=2,
            entropybits=30,
            strength=(0.333, 30)
        )

        passwords = {
            'qazwsx':           {'length', 'uppercase', 'numbers', 'special', 'nonletters', 'nonletterslc', 'entropybits', 'strength'},
            'qazwsxrfv':        {          'uppercase', 'numbers', 'special', 'nonletters', 'nonletterslc', 'entropybits', 'strength'},
            'qazwsxrfvTG':      {                       'numbers', 'special', 'nonletters',                                          },
            'qazwsxrfvTG94':    {                                  'special',                                                        },
            'qazwsxrfvTG94@$':  set(),
        }

        for password, expects in passwords.items():
            self.assertEqual(
                {t.name() for t in policy.test(password)},
                expects,
                'Testing {}'.format(password)
            )


password_strength/tests.py
""" These objects perform individual tests on a password, and report `True` of `False`. """

from .tests_base import ATest


class Length(ATest):
    """ Tests whether password length >= `length` """

    def __init__(self, length):
        super(Length, self).__init__(length)
        self.length = length

    def test(self, ps):
        return ps.length >= self.length


class Uppercase(ATest):
    """ Test whether the password has >= `count` uppercase characters """

    def __init__(self, count):
        super(Uppercase, self).__init__(count)
        self.count = count

    def test(self, ps):
        return ps.letters_uppercase >= self.count


class Numbers(Uppercase):
    """ Test whether the password has >= `count` numeric characters """

    def test(self, ps):
        return ps.numbers >= self.count


class Special(Uppercase):
    """ Test whether the password has >= `count` special characters """

    def test(self, ps):
        return ps.special_characters >= self.count


class NonLetters(Uppercase):
    """ Test whether the password has >= `count` non-letter characters """

    def test(self, ps):
        non_letters = ps.length - ps.letters
        return non_letters >= self.count


class NonLettersLc(Uppercase):
    """ Test whether the password has >= `count` non-lowercase characters """

    def test(self, ps):
        non_lowercase_letters = ps.length - ps.letters_lowercase
        return non_lowercase_letters >= self.count


class EntropyBits(ATest):
    """ Test whether the password has >= `bits` entropy bits.

    Entropy bits is the number of bits that is required to store the alphabet that's used in a password.
    It's a measure of how long is the alphabet.

    """

    def __init__(self, bits):
        super(EntropyBits, self).__init__(bits)
        self.bits = bits

    def test(self, ps):
        return ps.entropy_bits >= self.bits


class Strength(ATest):
    """ Test whether the password has >= `strength` strength.

        A password is evaluated to the strength of 0.333 when it has `weak_bits` entropy bits,
        which is considered to be a weak password. Strong passwords start at 0.666.
    """

    def __init__(self, strength, weak_bits=30):
        super(Strength, self).__init__(strength, weak_bits)
        self.strength = strength
        self.weak_bits = weak_bits

    def test(self, ps):
        return (1 - ps.weakness_factor) * ps.strength(self.weak_bits) >= self.strength


password_strength/policy.py
from .stats import PasswordStats
from . import tests as _tests


class PasswordPolicy(object):
    """ Perform tests on a password. """

    @classmethod
    def all_tests(cls):
        """ Get a dict of available tests

            :returns: { test-name: TestClass }
            :rtype: dict[type]
        """
        return dict(_tests.ATest.test_classes)

    @classmethod
    def from_names(cls, **tests):
        """ Init password policy from a dictionary of test definitions.

        A test definition is simply:

            { test-name: argument } or { test-name: [arguments] }

        Test name is just a lowercased class name.

        Example:

            PasswordPolicy.from_names(
                length=8,
                strength=(0.33, 30),
            )

        :param tests: Dict of test definitions.
        :rtype: PasswordPolicy
        :raises KeyError: wrong test name
        """
        _tests.ATest.test_classes['length'](8)
        tests = [ _tests.ATest.test_classes[name](
                      *(args if isinstance(args, (list, tuple)) else [args])
                  ) for name, args in tests.items() ]
        return cls(*tests)

    def __init__(self, *tests):
        """ Init password policy with a list of tests

        :param ts: List of tests to use for testing passwords
        :type ts: list[tests.ATest]
        """
        self._tests = tests

        assert all([isinstance(c, _tests.ATest) for c in tests]), 'Tests should be instances of password_strength.tests.ATest'

    def password(self, password):
        """ Get password stats bound to the tests declared in this policy.

        If in addition to tests you need to get statistics (e.g. strength) -- use this object to double calculations.

        See [`PasswordStats`](#passwordstats) for more details.

        :param password: Passphrase
        :type password: str|unicode
        :rtype: BoundPasswordStats
        """
        return BoundPasswordStats(password, self)

    def test(self, password):
        """ Perform tests on a password.

        Shortcut for: `PasswordPolicy.password(password).test()`.

        :param password: Passphrase
        :type password: str|unicode
        :return: List of tests that have failed
        :rtype: list[password_strength.tests.ATest]
        """
        return self.password(password).test()


class BoundPasswordStats(PasswordStats):
    """ PasswordStats bound to a PasswordPolicy """

    def __init__(self, password, policy):
        self._policy = policy
        super(BoundPasswordStats, self).__init__(password)

    def test(self):
        return super(BoundPasswordStats, self).test(self._policy._tests)


password_strength/tests_base.py
from six import with_metaclass

class ATestMeta(type):
    """ Metaclass that collects class names into `ATest.test_classes` dict.

        To define more classes, just subclass `ATest`.
        If class name starts with `_`, it's ignored.
    """

    def __new__(cls, name, bases, attrs):
        is_base = 'test_classes' in attrs
        test_classes = attrs['test_classes'] if is_base else ATest.test_classes

        cls = super(ATestMeta, cls).__new__(cls, name, bases, attrs)
        if not is_base and not name.startswith('_'):
            test_classes[name.lower()] = cls

        return cls


class ATest(with_metaclass(ATestMeta, object)):
    """ ATest is a base class for password tests.

        To create a custom test, just subclass it and implement the following methods:

        * __init__() that takes configuration arguments
        * test(ps) that tests a password, where `ps` is a `PasswordStats` object.
    """
    #: Test classes map: { name : class }
    test_classes = {}

    def __init__(self, *args):
        self.args = args  # Store args

    @classmethod
    def name(cls):
        """ Get test name """
        return cls.__name__.lower()

    def test(self, ps):
        """ Test a password with the test

        :param ps: Password stats
        :type ps: PasswordStats
        :return: Whether the test was passed
        :rtype: bool
        """
        raise NotImplementedError

    def __repr__(self):
        return '{cls}({args})'.format(cls=self.__class__.__name__, args=', '.join(map(str, self.args)))


password_strength/stats.py
from __future__ import division

import unicodedata
from collections import Counter
from math import log
import re
from functools import wraps
import sys, six

try:  # Python 2
    from itertools import izip as zip
except ImportError:
    pass  # Python 3


def cached_property(f):
    """ Property that will replace itself with a calculated value """
    name = '__' + f.__name__

    @wraps(f)
    def wrapper(self):
        if not hasattr(self, name):
            setattr(self, name, f(self))
        return getattr(self, name)
    return property(wrapper)


class PasswordStats(object):
    """ PasswordStats allows to calculate statistics on a password.

        It considers a password as a unicode string, and all statistics are unicode-based.
    """

    def __init__(self, password):
        self.password = six.text_type(password)

    #region Statistics

    @cached_property
    def alphabet(self):
        """ Get alphabet: set of used characters

        :rtype: set
        """
        return set(self.password)

    @cached_property
    def alphabet_cardinality(self):
        """ Get alphabet cardinality: alphabet length

        :rtype: int
        """
        return len(self.alphabet)

    @cached_property
    def char_categories_detailed(self):
        """ Character count per unicode category, detailed format.

        See: http://www.unicode.org/reports/tr44/#GC_Values_Table

        :returns: Counter( unicode-character-category: count )
        :rtype: collections.Counter
        """
        return Counter(map(unicodedata.category, self.password))

    @cached_property
    def char_categories(self):
        """ Character count per top-level category

        The following top-level categories are defined:

        - L: letter
        - M: Mark
        - N: Number
        - P: Punctuation
        - S: Symbol
        - Z: Separator
        - C: Other

        :return: Counter(unicode-character-category: count }
        :rtype: collections.Counter
        """
        c = Counter()
        for cat, n in self.char_categories_detailed.items():
            c[cat[0]] += n
        return c

    #endregion

    #region Counters

    @cached_property
    def length(self):
        """ Get password length

        :rtype: int
        """
        return len(self.password)

    @cached_property
    def letters(self):
        """ Count all letters

        :rtype: int
        """
        return self.char_categories['L']

    @cached_property
    def letters_uppercase(self):
        """ Count uppercase letters

        :rtype: int
        """
        return self.char_categories_detailed['Lu']

    @cached_property
    def letters_lowercase(self):
        """ Count lowercase letters

        :rtype: int
        """
        return self.char_categories_detailed['Ll']

    @cached_property
    def numbers(self):
        """ Count numbers

        :rtype: int
        """
        return self.char_categories['N']

    def count(self, *categories):
        """ Count characters of the specified classes only

        :param categories: Character categories to count
        :type categories: Iterable
        :rtype: int
        """
        return sum([int(cat_n[0] in categories) * cat_n[1] for cat_n in list(self.char_categories.items())])

    def count_except(self, *categories):
        """ Count characters of all classes except the specified ones

        :param categories: Character categories to exclude from count
        :type categories: Iterable
        :rtype: int
        """
        return sum([int(cat_n1[0] not in categories) * cat_n1[1] for cat_n1 in list(self.char_categories.items())])

    @cached_property
    def special_characters(self):
        """ Count special characters

        Special characters is everything that's not a letter or a number

        :rtype: int
        """
        return self.count_except('L', 'N')

    #region Security

    @cached_property
    def combinations(self):
        """ The number of possible combinations with the current alphabet

        :rtype: long
        """
        return self.alphabet_cardinality ** self.length

    @cached_property
    def entropy_bits(self):
        """ Get information entropy bits: log2 of the number of possible passwords

        https://en.wikipedia.org/wiki/Password_strength

        :rtype: float
        """
        return self.length * log(self.alphabet_cardinality, 2)

    @cached_property
    def entropy_density(self):
        """ Get information entropy density factor, ranged {0 .. 1}.

        This is ratio of entropy_bits() to max bits a password of this length could have.
        E.g. if all characters are unique -- then it's 1.0.
        If half of the characters are reused once -- then it's 0.5.

        :rtype: float
        """
        # Simplifying:
        #     entropy_bits / (length * log(length, 2)) =
        #   = log(alphabet_cardinality, 2) / log(length, 2) =
        #   = log(alphabet_cardinality, length)
        return log(self.alphabet_cardinality, self.length)

    def strength(self, weak_bits=30):
        """ Get password strength as a number normalized to range {0 .. 1}.

        Normalization is done in the following fashion:

        1. If entropy_bits <= weak_bits   -- linear in range{0.0 .. 0.33} (weak)
        2. If entropy_bits <= weak_bits*2 -- almost linear in range{0.33 .. 0.66} (medium)
        3. If entropy_bits > weak_bits*3  -- asymptotic towards 1.0 (strong)

        :param weak_bits: Minimum entropy bits a medium password should have.
        :type weak_bits: int
        :return: Normalized password strength:
            * <0.33 is WEAK
            * <0.66 is MEDIUM
            * >0.66 is STRONG
        :rtype: float
        """
        WEAK_MAX = 0.333333333

        if self.entropy_bits <= weak_bits:
            return WEAK_MAX * self.entropy_bits / weak_bits

        HARD_BITS = weak_bits*3
        HARD_VAL = 0.950

        # Here, we want a function that:
        # 1. f(x)=0.333 at x=weak_bits
        # 2. f(x)=0.950 at x=weak_bits*3 (great estimation for a perfect password)
        # 3. f(x) is almost linear in range{weak_bits .. weak_bits*2}: doubling the bits should double the strength
        # 4. f(x) has an asymptote of 1.0 (normalization)

        # First, the function:
        #       f(x) = 1 - (1-WEAK_MAX)*2^( -k*x)

        # Now, the equation:
        #       f(HARD_BITS) = HARD_VAL
        #       1 - (1-WEAK_MAX)*2^( -k*HARD_BITS) = HARD_VAL
        #                        2^( -k*HARD_BITS) = (1 - HARD_VAL) / (1-WEAK_MAX)
        #       k = -log2((1 - HARD_VAL) / (1-WEAK_MAX)) / HARD_BITS
        k = -log((1 - HARD_VAL) / (1-WEAK_MAX), 2) / HARD_BITS
        f = lambda x: 1 - (1-WEAK_MAX)*pow(2, -k*x)

        return f(self.entropy_bits - weak_bits)  # with offset

    #endregion

    #region Detectors

    _repeated_patterns_rex = re.compile(r'((.+?)\2+)', re.UNICODE | re.DOTALL | re.IGNORECASE)

    @cached_property
    def repeated_patterns_length(self):
        """ Detect and return the length of repeated patterns.

        You will probably be comparing it with the length of the password itself and ban if it's longer than 10%

        :rtype: int
        """
        length = 0
        for substring, pattern in self._repeated_patterns_rex.findall(self.password):
            length += len(substring)
        return length

    _sequences = (
        'abcdefghijklmnopqrstuvwxyz'  # Alphabet
        'qwertyuiopasdfghjklzxcvbnm'  # Keyboard
        '~!@#$%^&*()_+-='  # Keyboard special, top row
        '01234567890'  # Numbers
    )
    _sequences = _sequences + _sequences[::-1]  # reversed

    @cached_property
    def sequences_length(self):
        """ Detect and return the length of used sequences:

        - Alphabet letters: abcd...
        - Keyboard letters: qwerty, etc
        - Keyboard special characters in the top row: ~!@#$%^&*()_+
        - Numbers: 0123456

        :return: Total length of character sequences that are subsets of the common sequences
        :rtype: int
        """
        # FIXME: Optimize this. I'm sure there is a better way!...
        sequences_length = 0

        # Iterate through the string, with manual variable (to allow skips)
        i = 0
        while i < len(self.password):
            # Slice (since we use it often)
            password = self.password[i:]

            # Iterate over sequences to find longest common prefix
            j = -1
            common_length = 1
            while True:
                # Detect the first match with the current character
                # A character may appear multiple times
                j = self._sequences.find(password[0], j+1)
                if j == -1:
                    break

                # Find the longest common prefix
                common_here = ''
                for a, b in zip(password, self._sequences[j:]):
                    if a != b: break
                    else: common_here += a

                # It it's longer than previous discoveries -- store it
                common_length = max(common_length, len(common_here))

            # Repeated sequence?
            if common_length > 2:
                sequences_length += common_length

            # Next: skip to the end of the detected sequence
            i += common_length

        return sequences_length

    @cached_property
    def weakness_factor(self):
        """ Get weakness factor as a float in range {0 .. 1}

        This detects the portion of the string that contains:
        * repeated patterns
        * sequences

        E.g. a value of 1.0 means the whole string is weak, and 0.5 means half of the string is weak.

        Typical usage:

        password_strength = (1 - weakness_factor) * strength

        :return: Weakness factor
        :rtype: float
        """
        return min(1.0, (self.repeated_patterns_length + self.sequences_length) / self.length)

    #endregion

    def test(self, tests):
        """ Test the password against a list of tests

        :param tests: Test to do
        :type tests: Iterable[password_strength.tests.ATest]
        :return: list of tests that have failed
        :rtype: list[tests.ATest]
        """
        return [t
                for t in tests
                if not t.test(self)]
